File: backend/app/repositories/company_repository.py
"""
Company repository matching Java CompanyRepository interface.
"""
from typing import Optional
from sqlalchemy import func
from sqlalchemy.orm import Session

from app.models.company import Company


class CompanyRepository:
    """Data access layer for Company entity matching Java CompanyRepository."""
    
    def __init__(self, db: Session):
        self.db = db
    
    def find_by_name_ignore_case(self, name: str) -> Optional[Company]:
        """Matches Java: Optional<Company> findByNameIgnoreCase(String name)"""
        return (
            self.db.query(Company)
            .filter(func.lower(Company.name) == name.lower())
            .first()
        )
    
    def exists_by_name_ignore_case(self, name: str) -> bool:
        """Matches Java: boolean existsByNameIgnoreCase(String name)"""
        return self.db.query(
            self.db.query(Company)
            .filter(func.lower(Company.name) == name.lower())
            .exists()
        ).scalar()
    
    def save(self, company: Company) -> Company:
        """Save or update a company."""
        self.db.add(company)
        self.db.flush()
        self.db.refresh(company)
        return company
    
    def find_all(self, page: int, size: int) -> tuple[list[Company], int]:
        """
        Find all companies with pagination.
        Returns (companies, total_count)
        """
        query = self.db.query(Company)
        total = query.count()
        companies = query.offset(page * size).limit(size).all()
        return companies, total
    
    def count_all(self) -> int:
        """Count all companies."""
        return self.db.query(Company).count()

File: backend/src/main/java/com/codeforge/repository/CompanyRepository.java
package com.codeforge.repository;

import com.codeforge.entity.Company;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.Optional;

/**
 * Repository for Company entity
 */
@Repository
public interface CompanyRepository extends JpaRepository<Company, Long> {
    Optional<Company> findByName(String name);
    
    boolean existsByName(String name);
    
    Page<Company> findAll(Pageable pageable);
}
